### PR TITLE
DD Core: Use unique pointers in Store

### PR DESCRIPTION
**Pull request by ianna**
* DD Store migration to unique pointers

All compiles and the unit tests run ok.

@Dr15Jones and @cvuosalo please, review and comment. Thanks.

### CODE REVIEW

**Comment by ianna** [comment]
hold

**Comment by Dr15Jones** [comment]
@ianna given that you closed this pull request, do you still want us to look at it?

**Comment by ianna** [comment]
@Dr15Jones - yes, please. I'll reopen it when the rest is done. 
I just do not want to overload Fabio over the weekend. Thanks.

File: DetectorDescription/Core/src/DDCompactView.cc [hunk 1]
```diff
@@ -117,23 +117,24 @@ DDCompactView::DDCompactView()
...
-
-  // 2010-01-27 memory patch
-  // not sure this will stay, but for now we want to explicitely lock the global stores.
-  // lock the global stores.
-  DDMaterial::StoreT::instance().setReadOnly(false);
-  DDSolid::StoreT::instance().setReadOnly(false);
-  DDLogicalPart::StoreT::instance().setReadOnly(false);
-  DDSpecifics::StoreT::instance().setReadOnly(false);
-  DDRotation::StoreT::instance().setReadOnly(false);
+  // FIXME: The store is not read-only
```
[the remark this answers is not in the document]
Comment (ianna) [reply DetectorDescription/Core/src/DDCompactView.cc]: The store's instances read-only flag is set to false by default.
Ideally, the stores should be locked after the geometry has been parsed. In reality, it never happens. The magnetic field geometry producer needs a write access to the stores. Then, the CTPPS geometry producer needs a write access to the stores as well.

File: DetectorDescription/Core/interface/rep_type.h [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
   
   template< class N, class I > 
     struct rep_traits
     {
       using name_type = N;
-      using value_type = typename I::value_type;
+      //using value_type = typename I::value_type;
       using pointer = typename I::pointer;
-      using reference = typename I::reference;
+      using reference = I&;
```
[the remark this answers is not in the document]
Comment (ianna) [reply DetectorDescription/Core/interface/rep_type.h]: Thanks, good point!

**Comment by Dr15Jones** [comment]
@ianna did you do accidently merge CMSSW_10_2 versions into this branch?

**Comment by ianna** [comment]
unhold

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
code-checks

File: DetectorDescription/Core/plugins/DDAngular.cc [hunk 3]
```diff
@@ -155,8 +156,9 @@ DDAngular::execute( DDCompactView& cpv )
 				<< CONVERT_TO( phix, deg ) << ", 90.," 
 				<< CONVERT_TO( phiy, deg ) << ", 0, 0";
 	
-      rotation = DDrot( DDName( rotstr, m_idNameSpace ), new DDRotationMatrix(( *DDcreateRotationMatrix( theta, phix, theta, phiy,
-													 0., 0. ) * m_solidRot ))); 
+      rotation = DDrot( DDName( rotstr, m_idNameSpace ),
+			std::unique_ptr< DDRotationMatrix >( new DDRotationMatrix( *DDcreateRotationMatrix( theta, phix, theta, phiy,
```
[the remark this answers is not in the document]
Comment (ianna) [reply DetectorDescription/Core/plugins/DDAngular.cc]: Root complains profusely on ```std::make_unique```

[the remark this answers is not in the document]
Comment (Dr15Jones) [reply DetectorDescription/Core/plugins/DDAngular.cc]: What message does ROOT output?

**Comment by ianna** [comment]
@mrodozov - How can I re-trigger code checks? It looks like my fix was done too soon - before the final result came.

**Comment by Dr15Jones** [comment]
code-checks

**Comment by ianna** [comment]
+1

File: DetectorDescription/Core/interface/Store.h [hunk 4]
```diff
@@ -9,136 +9,129 @@
...
+    typename Store<N,I,K>::prep_type 
+    Store<N,I,K>::create( const name_type & n )
+    {
       prep_type tmp = nullptr;
-      auto result = reg_.insert(std::make_pair(n,tmp));
-      if (result.second) {
-         if (readOnly_) throw cms::Exception("DetectorDescriptionStore")<<" Store has been locked.  Illegal attempt to add " << n << " to a global store."; 
-         // ELSE     
-         result.first->second = new Rep_type(n,(I)nullptr);
+      auto result = reg_.insert( std::make_pair( n, tmp ));
```
Comment (Dr15Jones) [inline DetectorDescription/Core/interface/Store.h]: A more efficient way is to do 
```
auto result = reg.emplace( n, tmp );
```
this avoids the need to create a temporary pair.

File: DetectorDescription/Core/interface/Store.h [hunk 5]
```diff
@@ -9,136 +9,129 @@
...
+	throw cms::Exception( "DetectorDescriptionStore" )
+	  << " Store has been locked. Illegal attempt to add "
+	  << n << " to a global store."; 
+      // ELSE
       prep_type tmp = nullptr;
-      auto result = reg_.insert(std::make_pair(n,tmp));
-      if (!result.second) {
-         delete result.first->second->second;
-         result.first->second->second = p;
+      auto result = reg_.insert( std::make_pair( n, tmp ));
```
Comment (Dr15Jones) [inline DetectorDescription/Core/interface/Store.h]: Also use `emplace` here.

**Review by Dr15Jones — COMMENTED** [review]
Much better than what we had originally.

**Comment by ianna** [comment]
Thanks, @Dr15Jones

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by ianna** [comment]
@kpedro88 @alja @civanch - ping

**Comment by alja** [comment]
+1

**Comment by ianna** [comment]
@kpedro88 @civanch - please, check the PR

**Comment by civanch** [comment]
+1

File: DetectorDescription/Core/interface/Store.h [hunk 6]
```diff
@@ -9,136 +9,129 @@
...
-         // ELSE     
-         result.first->second = new Rep_type(n,(I)nullptr);
+      auto result = reg_.emplace( n, tmp );
+      if( result.second ) {
+	if( readOnly_ )
+	  throw cms::Exception( "DetectorDescriptionStore" )
+	    << " Store has been locked. Illegal attempt to add "
+	    << n << " to a global store."; 
+	// ELSE     
+	result.first->second = new Rep_type( n, nullptr );
```
Comment (kpedro88) [inline DetectorDescription/Core/interface/Store.h]: can smart pointers be used here as well?

File: DetectorDescription/Core/test/DDFilter.cppunit.cc [hunk 7]
```diff
@@ -120,7 +120,7 @@ void testDDFilter::checkFilters() {
         DDSpecifics ds{"EMinusVolume", {"//End[0]"}, values}; 
       } 
 
-      const DDRotation kXFlip = DDrot("xflip", new DDRotationMatrix{ ROOT::Math::RotationX{ kPI} } ); 
+      const DDRotation kXFlip = DDrot("xflip", std::unique_ptr< DDRotationMatrix >( new DDRotationMatrix{ ROOT::Math::RotationX{ kPI} } ));
```
Comment (kpedro88) [inline DetectorDescription/Core/test/DDFilter.cppunit.cc]: better to use `make_unique`

File: DetectorDescription/Core/plugins/DDLinear.cc [hunk 8]
```diff
@@ -104,7 +105,7 @@ DDLinear::execute( DDCompactView& cpv )
     LogDebug( "DDAlgorithm" ) << "DDLinear: Creating a new "
 			      << "rotation: IdentityRotation for " << ddname;
 	
-    rotation = DDrot( "IdentityRotation", new DDRotationMatrix());
+    rotation = DDrot( "IdentityRotation", std::unique_ptr< DDRotationMatrix >( new DDRotationMatrix()));
```
Comment (kpedro88) [inline DetectorDescription/Core/plugins/DDLinear.cc]: better to use `make_unique`

File: DetectorDescription/Core/src/DDCompactView.cc [hunk 9]
```diff
@@ -34,23 +34,17 @@ class DDDivision;
    \todo define a stable interface for navigation (don't expose the user to the graph!)
 */    
 // 
-DDCompactView::DDCompactView(const DDLogicalPart & rootnodedata)
+DDCompactView::DDCompactView( const DDLogicalPart & rootnodedata )
   : rep_( new DDCompactViewImpl( rootnodedata )),
```
Comment (kpedro88) [inline DetectorDescription/Core/src/DDCompactView.cc]: better to use `make_unique`

File: DetectorDescription/Core/src/DDCompactView.cc [hunk 10]
```diff
@@ -114,26 +111,28 @@ void DDCompactView::swap( DDCompactView& repToSwap ) {
 DDCompactView::DDCompactView()
   : rep_(new DDCompactViewImpl),
```
Comment (kpedro88) [inline DetectorDescription/Core/src/DDCompactView.cc]: better to use `make_unique`

File: DetectorDescription/Core/interface/Store.h [hunk 11]
```diff
@@ -9,136 +9,129 @@
...
-   void Store<N, I, K>::swap ( Store<N, I, K>& storeToSwap ) {
-      reg_.swap(storeToSwap.reg_);
-      storeToSwap.readOnly_ = readOnly_;
-   }
-   
+    }
+  
+  template<class N, class I, class K>
+    void Store<N, I, K>::swap ( Store<N, I, K>& storeToSwap ) {
+    reg_.swap(storeToSwap.reg_);
```
Comment (cvuosalo) [inline DetectorDescription/Core/interface/Store.h]: Can one swap values on a read-only object? Is that appropriate?

**Comment by ianna** [comment]
@kpedro88 - thanks, for your comments. I'll address them in a new PR. This one forces recompilation of pretty much everything. Besides there are two other PRs which depend on it. I'd rather had them integrated asis.

File: DetectorDescription/Core/src/DDCompactView.cc (hunk 9 above)
Comment (ianna) [reply DetectorDescription/Core/src/DDCompactView.cc]: The make_unique function is equivalent to: 
```
unique_ptr<T>(new T(std::forward<Args>(args)...))
```
This is a case where using make_unique will be more verbose then new. rep_ is a unique_ptr

File: DetectorDescription/Core/src/DDCompactView.cc (hunk 10 above)
Comment (ianna) [reply DetectorDescription/Core/src/DDCompactView.cc]: same here

File: DetectorDescription/Core/interface/Store.h (hunk 11 above)
Comment (ianna) [reply DetectorDescription/Core/interface/Store.h]: No. Unfortunately, the readOnly flag is always set to false and does not prevent internal modifications of the Store.

File: DetectorDescription/Core/interface/Store.h (hunk 6 above)
Comment (ianna) [reply DetectorDescription/Core/interface/Store.h]: No, Store needs to be redesigned.

File: DetectorDescription/Core/src/DDCompactView.cc (hunk 9 above)
Comment (kpedro88) [reply DetectorDescription/Core/src/DDCompactView.cc]: But it avoids the explicit use of `new`, which is preferable.

**Comment by kpedro88** [comment]
+upgrade

**Comment by kpedro88** [comment]
+1

**Pull request merged**